src/hotspot/share/code/icBuffer.cpp

```diff
@@ -247,26 +248,38 @@ void* InlineCacheBuffer::cached_value_for(CompiledIC *ic) {
 // Free CompiledICHolder*s that are no longer in use
 void InlineCacheBuffer::release_pending_icholders() {
   assert(SafepointSynchronize::is_at_safepoint(), "should only be called during a safepoint");
-  CompiledICHolder* holder = _pending_released;
-  _pending_released = nullptr;
+  CompiledICHolder* holder = Atomic::xchg(&_pending_released, (CompiledICHolder*)nullptr, memory_order_relaxed);
```

Comment:
I don't see why atomics are needed here.  Any GC threads calling queue_for_release() should be finished, and now we are single-threaded at a safepoint.

Reply:
Maybe file an RFE to do this work in parallel threads to reduce safepoint pause time?

Reply:
> I don't see why atomics are needed here. Any GC threads calling queue_for_release() should be finished, and now we are single-threaded at a safepoint.

I found the `xchg` more convenient than explicitly nulling out `_pending_release` in a separate statement. The implicit atomic op for `xchg` is just incidental (and won't have any impact on performance).

Will file an rfe about making this parallel.

Reply:
OK, a comment explaining that the atomic swap is convenient but not necessary would be helpful.

Reply:
Nevermind, it looks like you reverted the swap.